File: 23-networking/main.go
package main

import (
	"bufio"
	"fmt"
	"net"
	"strings"
	"time"
)

// ============================================================
// 示例 1: TCP Echo 服务器 + 客户端（自包含演示）
// ============================================================

func echoServerDemo() {
	fmt.Println("=== TCP Echo 服务器 ===")

	listener, err := net.Listen("tcp", "127.0.0.1:8080")
	if err != nil {
		fmt.Println("监听失败:", err)
		return
	}
	defer listener.Close()
	fmt.Println("服务器监听在 127.0.0.1:8080")

	// 在 goroutine 中启动客户端
	go func() {
		time.Sleep(100 * time.Millisecond)

		conn, err := net.Dial("tcp", "127.0.0.1:8080")
		if err != nil {
			fmt.Println("连接失败:", err)
			return
		}
		defer conn.Close()

		fmt.Println("客户端已连接")

		// 发送数据
		message := "Hello, Go Networking!\n"
		fmt.Fprintf(conn, message)
		fmt.Printf("发送: %s", message)

		// 读取回显
		reader := bufio.NewReader(conn)
		reply, _ := reader.ReadString('\n')
		fmt.Printf("收到回显: %s\n", reply)
	}()

	// 服务器接受一个连接
	conn, err := listener.Accept()
	if err != nil {
		fmt.Println("接受连接失败:", err)
		return
	}
	defer conn.Close()

	fmt.Printf("客户端连接: %s\n", conn.RemoteAddr())

	// 回显数据
	buf := make([]byte, 1024)
	n, _ := conn.Read(buf)
	fmt.Printf("服务端收到: %s", string(buf[:n]))
	conn.Write(buf[:n])

	time.Sleep(100 * time.Millisecond)
	fmt.Println()
}

// ============================================================
// 示例 2: 请求-响应服务器（简单协议）
// ============================================================

func handleConnection(conn net.Conn) {
	defer conn.Close()

	buf := make([]byte, 1024)
	n, err := conn.Read(buf)
	if err != nil {
		return
	}

	data := string(buf[:n])
	parts := strings.SplitN(data, ":", 2)

	if len(parts) != 2 {
		fmt.Fprintf(conn, "INVALID REQUEST\n")
		return
	}

	command, payload := parts[0], parts[1]
	fmt.Printf("命令: %s, 数据: %s\n", command, payload)

	switch command {
	case "PING":
		fmt.Fprintf(conn, "PONG\n")
	case "ECHO":
		fmt.Fprintf(conn, "%s", payload)
	default:
		fmt.Fprintf(conn, "UNKNOWN COMMAND\n")
	}
}

func requestResponseDemo() {
	fmt.Println("=== 请求-响应服务器 ===")

	listener, err := net.Listen("tcp", "127.0.0.1:8081")
	if err != nil {
		fmt.Println("监听失败:", err)
		return
	}
	defer listener.Close()
	fmt.Println("协议服务器监听在 127.0.0.1:8081")

	// 客户端在 goroutine 中发送请求
	go func() {
		time.Sleep(100 * time.Millisecond)

		// PING 请求
		conn, _ := net.Dial("tcp", "127.0.0.1:8081")
		fmt.Fprintf(conn, "PING:hello")
		buf := make([]byte, 1024)
		n, _ := conn.Read(buf)
		fmt.Printf("客户端收到: %s", string(buf[:n]))
		conn.Close()

		// ECHO 请求
		conn, _ = net.Dial("tcp", "127.0.0.1:8081")
		fmt.Fprintf(conn, "ECHO:Hello World!\n")
		n, _ = conn.Read(buf)
		fmt.Printf("客户端收到: %s", string(buf[:n]))
		conn.Close()
	}()

	// 处理 2 个请求
	for i := 0; i < 2; i++ {
		conn, err := listener.Accept()
		if err != nil {
			continue
		}
		handleConnection(conn)
	}

	time.Sleep(100 * time.Millisecond)
	fmt.Println("请求-响应演示完成")
	fmt.Println()
}

// ============================================================
// 示例 3: 多连接并发服务器
// ============================================================

func concurrentServerDemo() {
	fmt.Println("=== 并发服务器 ===")

	listener, err := net.Listen("tcp", "127.0.0.1:8082")
	if err != nil {
		fmt.Println("监听失败:", err)
		return
	}
	defer listener.Close()

	done := make(chan bool)

	// 启动多个并发客户端
	go func() {
		time.Sleep(100 * time.Millisecond)
		for i := 0; i < 4; i++ {
			go func(id int) {
				conn, _ := net.Dial("tcp", "127.0.0.1:8082")
				defer conn.Close()
				fmt.Fprintf(conn, "客户端 %d 的消息\n", id)
				buf := make([]byte, 1024)
				n, _ := conn.Read(buf)
				fmt.Printf("客户端 %d 收到: %s", id, string(buf[:n]))
			}(i)
		}
	}()

	// 服务器并发处理 4 个连接
	for i := 0; i < 4; i++ {
		conn, _ := listener.Accept()
		go func(c net.Conn) {
			defer c.Close()
			buf := make([]byte, 1024)
			n, _ := c.Read(buf)
			c.Write(buf[:n]) // 回显
		}(conn)
	}

	go func() {
		time.Sleep(500 * time.Millisecond)
		done <- true
	}()

	<-done
	fmt.Println("并发服务器演示完成")
	fmt.Println()
}

func main() {
	echoServerDemo()
	requestResponseDemo()
	concurrentServerDemo()
}

File: 23-networking/main.rs
use std::io::{Read, Write};
use std::net::{TcpListener, TcpStream};
use std::thread;
use std::time::Duration;

// ============================================================
// 示例 1: TCP Echo 服务器 + 客户端
// ============================================================

fn echo_server_demo() {
    println!("=== TCP Echo 服务器 ===");

    let listener = TcpListener::bind("127.0.0.1:8080").expect("绑定失败");
    println!("服务器监听在 127.0.0.1:8080");

    // 在子线程中启动客户端
    thread::spawn(|| {
        thread::sleep(Duration::from_millis(100));

        let mut stream = TcpStream::connect("127.0.0.1:8080").expect("连接失败");
        println!("客户端已连接");

        let message = b"Hello, Rust Networking!\n";
        stream.write_all(message).expect("发送失败");
        print!("发送: {}", String::from_utf8_lossy(message));

        let mut buf = [0u8; 1024];
        let n = stream.read(&mut buf).expect("读取失败");
        print!("收到回显: {}\n", String::from_utf8_lossy(&buf[..n]));
    });

    // 服务器接受一个连接
    let (mut stream, addr) = listener.accept().expect("接受连接失败");
    println!("客户端连接: {}", addr);

    let mut buf = [0u8; 1024];
    let n = stream.read(&mut buf).expect("读取失败");
    print!("服务端收到: {}", String::from_utf8_lossy(&buf[..n]));
    stream.write_all(&buf[..n]).expect("写入失败");

    thread::sleep(Duration::from_millis(100));
    println!();
}

// ============================================================
// 示例 2: 请求-响应服务器
// ============================================================

fn handle_connection(mut stream: TcpStream) {
    let mut buf = [0u8; 1024];
    let n = match stream.read(&mut buf) {
        Ok(n) => n,
        Err(_) => return,
    };

    let data = String::from_utf8_lossy(&buf[..n]);
    let parts: Vec<&str> = data.splitn(2, ':').collect();

    if parts.len() != 2 {
        let _ = stream.write_all(b"INVALID REQUEST\n");
        return;
    }

    let (command, payload) = (parts[0], parts[1]);
    println!("命令: {}, 数据: {}", command, payload);

    match command {
        "PING" => {
            let _ = stream.write_all(b"PONG\n");
        }
        "ECHO" => {
            let _ = stream.write_all(payload.as_bytes());
        }
        _ => {
            let _ = stream.write_all(b"UNKNOWN COMMAND\n");
        }
    }
}

fn request_response_demo() {
    println!("=== 请求-响应服务器 ===");

    let listener = TcpListener::bind("127.0.0.1:8081").expect("绑定失败");
    println!("协议服务器监听在 127.0.0.1:8081");

    // 客户端线程
    thread::spawn(|| {
        thread::sleep(Duration::from_millis(100));

        // PING 请求
        {
            let mut stream = TcpStream::connect("127.0.0.1:8081").unwrap();
            stream.write_all(b"PING:hello").unwrap();
            let mut buf = [0u8; 1024];
            let n = stream.read(&mut buf).unwrap();
            print!("客户端收到: {}", String::from_utf8_lossy(&buf[..n]));
        }

        // ECHO 请求
        {
            let mut stream = TcpStream::connect("127.0.0.1:8081").unwrap();
            stream.write_all(b"ECHO:Hello World!\n").unwrap();
            let mut buf = [0u8; 1024];
            let n = stream.read(&mut buf).unwrap();
            print!("客户端收到: {}", String::from_utf8_lossy(&buf[..n]));
        }
    });

    // 处理 2 个请求
    for _ in 0..2 {
        match listener.accept() {
            Ok((stream, _)) => handle_connection(stream),
            Err(_) => continue,
        }
    }

    thread::sleep(Duration::from_millis(100));
    println!("请求-响应演示完成");
    println!();
}

// ============================================================
// 示例 3: 多连接并发服务器
// ============================================================

fn concurrent_server_demo() {
    println!("=== 并发服务器 ===");

    let listener = TcpListener::bind("127.0.0.1:8082").expect("绑定失败");

    // 启动多个客户端线程
    thread::spawn(|| {
        thread::sleep(Duration::from_millis(100));

        let mut handles = vec![];
        for i in 0..4 {
            let handle = thread::spawn(move || {
                let mut stream = TcpStream::connect("127.0.0.1:8082").unwrap();
                let msg = format!("客户端 {} 的消息\n", i);
                stream.write_all(msg.as_bytes()).unwrap();

                let mut buf = [0u8; 1024];
                let n = stream.read(&mut buf).unwrap();
                print!("客户端 {} 收到: {}", i, String::from_utf8_lossy(&buf[..n]));
            });
            handles.push(handle);
        }

        for h in handles {
            h.join().unwrap();
        }
    });

    // 服务器用线程并发处理连接
    let mut handles = vec![];
    for _ in 0..4 {
        let (stream, _) = listener.accept().unwrap();
        let handle = thread::spawn(move || {
            let mut stream = stream;
            let mut buf = [0u8; 1024];
            let n = stream.read(&mut buf).unwrap();
            stream.write_all(&buf[..n]).unwrap();
        });
        handles.push(handle);
    }

    for h in handles {
        h.join().unwrap();
    }

    println!("并发服务器演示完成");
    println!();
}

fn main() {
    echo_server_demo();
    request_response_demo();
    concurrent_server_demo();
}
